Userland/Libraries/LibTLS/TLSv12.cpp

```diff
@@ -200,40 +204,101 @@ bool Context::verify_chain() const
         local_chain = &certificates;
     }
 
-    // FIXME: Actually verify the signature, instead of just checking the name.
-    HashMap<String, String> chain;
-    HashTable<String> roots;
-    // First, walk the root certs.
-    for (auto& cert : root_certificates) {
-        roots.set(cert.subject.subject);
-        chain.set(cert.subject.subject, cert.issuer.subject);
+    if (local_chain->is_empty()) {
+        dbgln("verify_chain: Attempting to verify an empty chain");
+        return false;
     }
 
-    // Then, walk the local certs.
-    for (auto& cert : *local_chain) {
-        auto& issuer_unique_name = cert.issuer.unit.is_empty() ? cert.issuer.subject : cert.issuer.unit;
-        chain.set(cert.subject.subject, issuer_unique_name);
-    }
+    for (size_t cert_index = 0; cert_index < local_chain->size(); ++cert_index) {
+        auto cert = local_chain->at(cert_index);
 
-    // Then verify the chain.
-    for (auto& it : chain) {
-        if (it.key == it.value) { // Allow self-signed certificates.
-            if (!roots.contains(it.key))
-                dbgln("Self-signed warning: Certificate for {} is self-signed", it.key);
-            continue;
-        }
+        auto subject_string = cert.subject_identifier_string();
+        auto issuer_string = cert.issuer_identifier_string();
 
-        auto ref = chain.get(it.value);
-        if (!ref.has_value()) {
-            dbgln("{}: Certificate for {} is not signed by anyone we trust ({})", this, it.key, it.value);
+        if (!cert.is_valid()) {
+            dbgln("verify_chain: Certificate is not valid {}", subject_string);
             return false;
         }
 
-        if (ref.value() == it.key) // Allow (but warn about) mutually recursively signed cert A <-> B.
-            dbgln("Co-dependency warning: Certificate for {} is issued by {}, which itself is issued by {}", ref.value(), it.key, ref.value());
+        auto maybe_root_certificate = root_certificates.get(issuer_string);
+        if (maybe_root_certificate.has_value()) {
+            auto root_certificate = maybe_root_certificate.release_value();
```

Comment:
This copies the certificate, instead:
```suggestion
            auto& root_certificate = *maybe_root_certificate;
```